### PR TITLE
ATB-1269 | Configure EventBridge Pipe to integrate TxMA target queue and AIS TxMaIngress Queue

### PR DESCRIPTION
**Pull request by chiarapascucci-gds**
## Proposed changes
This PR merges changes to integrate AIS with TxMA

### What changed
created new resources:
- event bridge pipe to pick up messages from the queue in the txma account (from staging onwards) and deliver those to AIS own TxMAIngress SQS Queue
- execution role and related policy for TxMAIngressPipe
- Log group for the pipe
- Queue and DLQ for dev and build environment to act as a stand-in for the actual TxMA queue, which we only integrate with in staging, integration, and production envs
- SSM parameter storing the name of the pipe to be used when creating alarms related to the pipe
Code changes in intervention handler to account for the new message structure 

### Why did it change
Integrating AIS with TxMA

### Issue tracking
- [ATB-1269](https://govukverify.atlassian.net/browse/ATB-1269)

## Testing
Added UTs 
Screenshot 2023-12-15 at 09 04 36

Deployed changes to dev env and sent messages on the `TxMAEmulatedSourceQueue` to simulate an event originating within txma
Sent block account --> unblock account --> suspend with psw and id reset --> reverify identity --> reset password ✅ 
sent message with invalid JSON in body, it is not retried ✅ 
Screenshot 2023-12-15 at 09 16 28
sent message slightly in the future (retried)
Screenshot 2023-12-15 at 09 20 02
Set logging level for pipe so not include execution data, but still being able to see errors (example)
Screenshot 2023-12-15 at 09 21 23

## Checklists
- [x] Did not commit any not-required changes to the `src/infra/**/samconfig.toml`
- [x] Tested changes and included test evidence in the PR, if appropriate
- [x] Included all required tags and other properties for any new resources in the SAM template
- [x] Ensured that any new resources in the SAM Template follow appropriate naming conventions
- [x] Ensured that naming of new resources is compatible with deploying multiple stacks with custom stack names during development
- [x] Ensured that no log lines include PII or other sensitive data
- [x] Implemented unit testing for any new logic implemented, if appropriate
- [x] Ensured that all commits in this PR are signed
- [x] Ensured appropriate code coverage is maintained by unit tests

[ATB-1269]: https://govukverify.atlassian.net/browse/ATB-1269?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/infra/main/template.yaml [hunk 1]
```diff
@@ -762,6 +784,187 @@ Resources:
       AliasName: !Sub "alias/${AWS::StackName}-SQSEncryptionKey"
       TargetKeyId: !GetAtt SQSEncryptionKey.Arn
 
+### TxMA queue integration ####
+  TxMaIngressPipe:
```
Comment (TaherKapasi) [inline src/infra/main/template.yaml:788]: TxMaIngressPipe -> TxMAIngressPipe - just to keep the same as other resources

File: src/infra/main/template.yaml [hunk 2]
```diff
@@ -762,6 +784,187 @@ Resources:
...
+        Environment: !Ref Environment
+        Product: !Ref ProductTagValue
+        System: !Ref SystemTagValue
+        Owner: !Ref OwnerTagValue
+        Source: !Ref SourceTagValue
+
+  TxMAIngressPipeRole:
+    Type: AWS::IAM::Role
+    Properties:
+      RoleName: !Sub '${AWS::StackName}-TxMaIngressPipe'
```
Comment (TaherKapasi) [inline src/infra/main/template.yaml:817]: same as above

File: src/infra/main/template.yaml [hunk 3]
```diff
@@ -762,6 +784,187 @@ Resources:
...
+            Action:
+              - "sqs:SendMessage"
+            Resource: !GetAtt TxMAIngressQueue.Arn #to publish messages to target queue)
+      Roles:
+        - !Ref TxMAIngressPipeRole
+
+  TxMAIngressPipeLogGroup:
+    Type: AWS::Logs::LogGroup
+    Properties:
+      LogGroupName: !Sub "/aws/pipe/${AWS::StackName}-TxMaIngressPipe"
```
Comment (TaherKapasi) [inline src/infra/main/template.yaml:886]: same as above

File: src/infra/main/template.yaml [hunk 4]
```diff
@@ -762,6 +784,187 @@ Resources:
...
+        - Key: System
+          Value: !Ref SystemTagValue
+        - Key: Environment
+          Value: !Ref Environment
+        - Key: Owner
+          Value: !Ref OwnerTagValue
+        - Key: Source
+          Value: !Ref SourceTagValue
+        - Key: Name
+          Value: !Sub "/aws/pipe/${AWS::StackName}-TxMaIngressPipe"
```
Comment (TaherKapasi) [inline src/infra/main/template.yaml:901]: same

File: src/handlers/tests/interventions-processor-lambda.test.ts [hunk 5]
```diff
@@ -405,5 +464,89 @@ describe('intervention processor handler', () => {
       expect(logAndPublishMetric).toHaveBeenCalledWith('CONFIDENCE_LEVEL_TOO_LOW');
       expect(logger.warn).toHaveBeenCalledWith('Received interventions has low level of confidence: P1');
     });
+
+    it('It should log an error and continue if the inner message body is not valid JSON', async () => {
```
Comment (liamquayle) [inline src/handlers/tests/interventions-processor-lambda.test.ts:468]: absolute nit pick, apols:
```suggestion
    it('should log an error and continue, if the inner message body is not valid JSON', async () => {
```

File: src/infra/main/template.yaml [hunk 6]
```diff
@@ -1082,3 +1284,8 @@ Outputs:
   SqsQueueUrl:
```
Comment (TaherKapasi) [inline src/infra/main/template.yaml:1284]: Could we rename the to TxMAEgressSQSQueueUrl

File: src/handlers/tests/interventions-processor-lambda.test.ts [hunk 7]
```diff
@@ -405,5 +464,89 @@ describe('intervention processor handler', () => {
...
+      };
+      expect(await handler({ Records: [mockRecord] }, mockContext)).toEqual({
+        batchItemFailures: [],
+      });
+      expect(logger.error).toHaveBeenCalledWith('Message body is not valid JSON.');
+      expect(logAndPublishMetric).toHaveBeenCalledWith(MetricNames.INVALID_EVENT_RECEIVED);
+
+    });
+
+    it('If a valid and invalid events are received none should be retried', async () => {
```
Comment (liamquayle) [inline src/handlers/tests/interventions-processor-lambda.test.ts:493]:
```suggestion
    it('should be not retry, if both valid and invalid events are received', async () => {
```

File: src/infra/main/template.yaml [hunk 8]
```diff
@@ -762,6 +784,187 @@ Resources:
...
+      PolicyName: !Sub "${AWS::StackName}-TxMAIngressPipePolicy"
+      PolicyDocument:
+        Version: "2012-10-17"
+        Statement:
+          - Effect: Allow
+            Action:
+              - "logs:CreateLogStream"
+              - "logs:PutLogEvents"
+            Resource:
+              - !GetAtt TxMAIngressPipeLogGroup.Arn #to create own log stream
```
Comment (liamquayle) [inline src/infra/main/template.yaml:853]: do we need these comments?

**Review by TaherKapasi — APPROVED** [review]

**Review by liamquayle — APPROVED** [review]

**Review by KG700 — APPROVED** [review]

File: src/infra/main/template.yaml (hunk 6 above)
Comment (chiarapascucci-gds) [reply src/infra/main/template.yaml:1284]: Done - I checked alarm and core template and that value is not imported there, so no further changes as far as I could see. But please let me know if you that is imported somewhere and we need to update

File: src/handlers/tests/interventions-processor-lambda.test.ts (hunk 7 above)
Comment (chiarapascucci-gds) [reply src/handlers/tests/interventions-processor-lambda.test.ts:493]: sorry, I don't think the suggestions reads correctly

**Pull request merged**